### DISCUSSION TITLE
Curly braces in class declarations

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#5929] [Language Ideas]
Curly braces can be used arbitrarily in functions. And the variables declared in curly braces are not visible to the outside  
```
{ int num = 0; }  
Console.WriteLine(num);//CS103
```
I propose to allow this functionality in class declarations.
If the member in curly braces does not have access modifier.
Then it is invisible to the outside.

for exampl:
https://github.com/dotnet/roslyn/issues/850#issue-58855325
you can:
```
{  
    string field1;  
    prive string filed2;  
    public string Filed{get;set;}  
}  
public void show()  
{  
    Console.WriteLine(field1);//not allow  
    Console.WriteLine(field2);//allow  
    Console.WriteLine(Field);//allow  
}  
```
Nested is also allowed

**Comment by Joe4evr** [comment]
But for what purpose? Would your scenario be solved by any of #133, #140 (which is being done and [in preview *right now*](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLTIgWwJaqnGAe1QAcFCTkAaGEANwgWzAE8AfAAQAYACDgRgB0AYUKYShAHYRJMALKEAJhAA2AbgCwAKA4BmPgCYewniB4BJAHKEYzFgAVyJRjBbCAFlEkBzCIu0A3to8IXz6ArwAysGhQVqhCTy+MDwAvAB8PGDYqopqPAD0BQDGiEiyKiw80sUoyFAILFQ8AFZwyClQPADq5vY8AI5w2AgA1jGJPBgpGTyREDAAgpKK1rasABRIYFk5KorNdFAqcBDNklCYEIRgG5EAlPea8aEAvtoTYTwQDLI8jhQXG5PD4/ABRX4wAASXkUKkY/ycQI8Xl8eQ+LxCZGwRxgED4ABY5gtlqsbHYADwAFXSWwgOypWX2zUZRxOZz4/F4ZAolkuEHunzikxCYH2aR4bNOzxFiMBCFcKNBigA/IJzJI6IRRhANjB3NhqNUIAB3DY85wKlh8q5mC02gVPT7vLSvIA==)), or #5529?

**Reply by jnm2** [reply]
Ooh! I didn't know it was in preview. [(You can use `CallerMemberName` to good effect here too.)](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLTIgWwJaqnGAe1QAcFCTkAaGEANwgWzAE8AfAAQAYACDgRgB0AYUKYShAHYRJMALKEAJhAA2AbgCwAKG58hAJTizsmCCLElsKxgGVGdbAGMUmrdoDEkuCpVRg1nhk/a21tDgBmPgAmHmEeEB4AeWAMBDpgiGSAKwhHGG0Ab20eEr5IgV4bHgKeAHMIGB4AXgA+HjBsVUU1HgxG1p4AFQQWOxgAQUlFADlCGGYWAAokMHbOlUUqHnSVOAgASh6AX20Ttx1Iv2QYBCg86KSU+wzs3MaEgElZ+dYABXISIwYCxhAALKCSeqKQrFUoRQIMWQ8f4UIEg8GQiCKACiiJgAAkIYprAgAPzIgFosEQqGuUo8WElMhzN5YnjAQiEFRDEZjSYzOYLAA8gxaywgq0Gay6WylOz2WwA2sIoD5GHIsMBGNMoKYALp6Ljk5mAhDAnWmZo8Lw+faM6r2+nMHiLQzGUz41Sm5CCbEARzgquQiw6Mu2qr2+32fAA7O0gxA6fTSqGNlb5YnHaVEpIUabgdTMYpFia0RaDknk7GeDc9pWeGd6fbmTBWYo+AAWJK5ylm9E0rGLZWqkkazBahDlg0VY29826iBWm0qO1aelFNdVimovuFqGkwQfSR0QgAawgixgoOw1GtEAA7iW5yxy1H60cm1ozkA)

**Reply by 333fred** [reply]
It is not in preview. It is in a feature branch. Still work to be done.

**Reply by zms9110750** [reply]
Because it is more free, it can not only be used for semi-automatic attributes
```csharp
	readonly Queue<GameObject> pool = new Queue<GameObject>();
	GameObject GetGameBojectFromFile() => default;
	public GameObject GameObjectPool
	{
		get
		{
			GameObject game;
			if (pool.Count == 0)
				game = GetGameBojectFromFile();
			else
				game = pool.Dequeue();
			game.activation = true;
			return game;
		}
		set
		{
			value.activation = false;
			pool.Enqueue(value);
		}
	}
```
I don't want the queue of the object pool to be accessed by others.
In addition, there may be multiple object pools, which share the method of obtaining from files.
But this method does not want to be accessed by others

**Reply by theunrepentantgeek** [reply]
Putting aside the fact this property is doing way _way **way**_ too much, what you're seeing is evidence there's another class yearning to be discovered:

``` csharp
private class GameObjectPool
{
    private readonly Queue<GameObject> _pool = new()
    public GameObject Pull()
    {
		GameObject game;
		if (pool.Count == 0)
			game = GetGameBojectFromFile();
		else
			game = pool.Dequeue();
		game.activation = true;
		return game;        
    }

    public void Push(GameObject game)
	{
		game.activation = false;
		pool.Enqueue(game);
	}
}
```

C# already has good mechanisms for encapsulation of functionality and restricting scope of access. This is literally what **class** and **struct** declarations are for.

**Comment by zms9110750** [comment]
Compared to #133 ,You can restrict fields to be accessed only through certain (multiple) methods, not necessarily get / set.
Compared to #140,the restricted fields that can be generated automatically are different in kind / quantity for their attributes

I remember seeing a problem in a discussion of this kind of problem. Whether the constructor should have access to qualified fields inside the property.This approach preserves two options. Because he's more free.
The whole process is similar to writing local methods in curly braces,If you write a local method in for / if, it cannot be accessed outside of for / if.Extending this notation to class declarations can solve some problems
